### PR TITLE
Add native Permuto and multilevel LoD encodings

### PR DESCRIPTION
**Pull request by shsolanki-1**
## Summary

- Add native Permuto training, parameter-gradient, input-gradient, and supported double-backward paths.
- Generalize Permuto construction across input dimensions `1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 16, 24` and features per level `1, 2, 4, 8`.
- Generalize `MultiLevelEncodingLoD` to wrap multilevel encodings with hard or soft level selection.
- Preserve base layouts, matrix strides, parameter ownership, CUDA streams, gradient modes, empty batches, and serialized configuration.
- Document the public configuration, execution, and derivative contracts.

## Motivation

The existing Permuto factory exposes one fixed specialization, and the existing level-of-detail wrapper is tied to Permuto and hard level selection. This change exposes the existing native implementation over a defined configuration matrix and provides one native multilevel LoD wrapper for Permuto and Grid encodings.

## Compatibility

The existing five-dimensional configuration with two features per level retains its parameter order, lattice tables, outputs, gradients, and serialized hyperparameters. `n_features` and `n_grid_features` are accepted as total-feature aliases and serialize to canonical `n_levels` and `n_features_per_level` fields.

The LoD ratio is scheduler state. Its input gradient remains zero. Nested LoD wrappers are rejected because their mutable level state does not compose consistently.

## Validation

- Configured and built Release tests with CUDA 12.8, GCC 12.3, and `TCNN_CUDA_ARCHITECTURES=89`.
- Passed `test_permuto` with 29 test cases and 425,577 assertions on an RTX 4090.
- Passed the focused non-JIT `FullyFusedMLP` double-backward test with 141,383 assertions.
- Verified the PR range with `git diff --check`.

## Build Impact

The complete native matrix creates 104 Permuto specializations: 13 input dimensions, four feature widths, and two scalar types. In the measured Release build, clean build time increased by 1.2155x, `encoding.cu.o` increased by 2.6306x, and `libtiny-cuda-nn.a` increased by 1.6015x. The retained five-dimensional, two-feature specialization kept the same generated device-code size.

Generated inference is intentionally excluded from this PR and is provided by stacked follow-up #539. Generated backward and generated double backward remain unsupported.

### CODE REVIEW

File: include/tiny-cuda-nn/encodings/grid.h [hunk 1]
```diff
@@ -239,7 +239,7 @@ __global__ void kernel_grid_backward(
 		max_level = (max_level * num_grid_features) / N_FEATURES_PER_LEVEL;
 	}
 
-	if (level > max_level + 1e-3f) {
+	if (level >= max_level + 1e-3f) {
```
Comment (njroussel) [inline include/tiny-cuda-nn/encodings/grid.h]: Is this change for consistency with the same check if `kernel_grid` (forward) ? Or is there some other reason for it too?

File: include/tiny-cuda-nn/network_with_input_encoding.h [hunk 2]
```diff
@@ -72,10 +76,25 @@ class NetworkWithInputEncoding : public Network<float, T> {
...
 
 		forward->network_input = GPUMatrixDynamic<T>{m_encoding->padded_output_width(), input.n(), stream, m_encoding->preferred_output_layout()};
 		forward->encoding_ctx = m_encoding->forward(stream, input, &forward->network_input, use_inference_params, prepare_input_gradients);
 		forward->network_ctx = m_network->forward(stream, forward->network_input, output, use_inference_params, true);
+		if (output && prepare_input_gradients) {
+			forward->network_output = GPUMatrixDynamic<T>{output->m(), output->n(), stream, output->layout()};
+			if (output->layout() == AoS) {
+				parallel_for_gpu_aos(stream, output->n(), output->m(), [src = output->view(), dst = forward->network_output.view()] __device__(size_t i, size_t j) {
+					dst((uint32_t)i, (uint32_t)j) = src((uint32_t)i, (uint32_t)j);
+				});
```
Comment (njroussel) [inline include/tiny-cuda-nn/network_with_input_encoding.h]:
```cpp
parallel_for_gpu_aos(stream, output->n(), output->m(), [src = output->view(), dst = forward->network_output.view()] __device__(size_t elem, size_t dim) {
        dst((uint32_t)dim, (uint32_t)elem) = src((uint32_t)dim, (uint32_t)elem);
});
```
Pretty sure the indexing is flipped here.

File: include/tiny-cuda-nn/network_with_input_encoding.h [hunk 3]
```diff
@@ -72,10 +76,25 @@ class NetworkWithInputEncoding : public Network<float, T> {
...
+			} else {
+				parallel_for_gpu(stream, output->n() * output->m(), [src = output->data(), dst = forward->network_output.data()] __device__(size_t i) {
+					dst[i] = src[i];
```
Comment (njroussel) [inline include/tiny-cuda-nn/network_with_input_encoding.h]: I'm not convinced we can assume that everything is continguous here, could it not be strided ?

File: include/tiny-cuda-nn/network_with_input_encoding.h [hunk 4]
```diff
@@ -72,10 +76,25 @@ class NetworkWithInputEncoding : public Network<float, T> {
...
 
 		auto forward = std::make_unique<ForwardContext>();
+		if (batch_size == 0) {
+			return forward;
+		}
 
 		forward->network_input = GPUMatrixDynamic<T>{m_encoding->padded_output_width(), input.n(), stream, m_encoding->preferred_output_layout()};
 		forward->encoding_ctx = m_encoding->forward(stream, input, &forward->network_input, use_inference_params, prepare_input_gradients);
 		forward->network_ctx = m_network->forward(stream, forward->network_input, output, use_inference_params, true);
+		if (output && prepare_input_gradients) {
```
Comment (njroussel) [inline include/tiny-cuda-nn/network_with_input_encoding.h]: This seems a bit too general, no? We'd almost always copy the output, not just for cases that want a double backward.
